Project-Hotpot/Project-Hotpot/ViewControllers/HomeViewController.swift

```diff
@@ -83,6 +84,41 @@ class HomeViewController: UIViewController {
     }
     
     // MARK: - helper functions
+    func addToPlaylist(){
+        //extract pfsong object
+        //TODO: consider using optional completion blocks maybe to avoid using repeat code?
+        //TODO: Ask Toby for style advice on this function especially -> nested functions with completion handlers?
```

Comment:
Pyramids of doom like this are a big drawback of the callback pattern. My best suggestion would be to define each of these closures as siblings of `addToPlaylist`, and then have each method call the next one it needs. This should mean each method only needs to go one-tab deep, instead of continuing to tab over in the same parent method.

Again, the long-term fix to this is the move to async/await (https://docs.swift.org/swift-book/LanguageGuide/Concurrency.html), which is going to be fantastic.

But I have one other thought; are you certain you need all of these calls? Or is it possible you can use the device's cache to save one of these calls (i.e. is it truly necessary to fetch the playlist every time you want to save another song?)